### PR TITLE
Enable EnforceCodeStyleInBuild by default

### PR DESCRIPTION
**Pull request by jgonz120**
## Bug

Fixes:

Regression? Last working version:

## Description

Updated the common.project.props to properly default the EnforceCodeStyleInBuild property to "true". Fixed build errors that arose from this

## PR Checklist

- [X] PR has a meaningful title
- [ ] PR has a linked issue.
- [X] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [X] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [X] N/A

### CODE REVIEW

File: build/common.project.props [hunk 1]
```diff
@@ -124,13 +124,17 @@
     <MinimumVisualStudioVersion>15.0</MinimumVisualStudioVersion>
     <VsixOutputDirName>VS15</VsixOutputDirName>
     <CodeAnalysisRuleSet Condition="'$(CodeAnalysisRuleSet)' == ''">$(BuildCommonDirectory)NuGet.ruleset</CodeAnalysisRuleSet>
-    <EnforceCodeStyleInBuild Condition=" '$(EnforceCodeStyleInBuild)' != '' " >true</EnforceCodeStyleInBuild>
+    <WarningsNotAsErrors>EnableGenerateDocumentationFile</WarningsNotAsErrors>
```
Comment (nkolev92) [inline build/common.project.props]:
```suggestion
    <WarningsNotAsErrors>$(WarningsNotAsErrors);EnableGenerateDocumentationFile</WarningsNotAsErrors>
```

We should add not overwrite.

fwiw, I had no clue there's a warning identified as  `EnableGenerateDocumentationFile` and that it can be suppressed like that :D

File: build/common.project.props [hunk 2]
```diff
@@ -124,13 +124,17 @@
...
   </PropertyGroup>
 
   <PropertyGroup>
     <VsixPublishDestination>$(ArtifactRoot)$(VsixOutputDirName)\</VsixPublishDestination>
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(EnforceCodeStyleInBuild)' != 'false'">
```
Comment (nkolev92) [inline build/common.project.props]: `== ''` should have been the pattern. 

Any reason not to update that on line 127?

Many properties are defined there.

Comment (jgonz120) [reply build/common.project.props]: The documentation on conditions shows that only certain elements support the condition statement.  https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-conditions?view=vs-2022#supported-elements

File: build/common.project.props (hunk 1 above)
Comment (jgonz120) [reply build/common.project.props]: I wasn't sure it would work tbh xD

File: build/common.project.props (hunk 2 above)
Comment (jgonz120) [reply build/common.project.props]: But ya, switched it back to == ''

Comment (jgonz120) [reply build/common.project.props]: Oh I guess "Property" counts as any property? whatever lol

File: build/TestShared/GlobalSuppressions.cs [hunk 3]
```diff
@@ -5,17 +5,3 @@
 // attributes that are applied to this project.
 // Project-level suppressions either have no target or are given 
 // a specific target and scoped to a namespace, type, member, etc.
```
Comment (zivkan) [inline build/TestShared/GlobalSuppressions.cs]: This file is now effectively empty (it only has comments). You should delete it.

Comment (nkolev92) [reply build/TestShared/GlobalSuppressions.cs]: Makes sense, we should delete this.

**Review by nkolev92 — DISMISSED** [review]

File: build/common.project.props [hunk 4]
```diff
@@ -124,7 +124,8 @@
     <MinimumVisualStudioVersion>15.0</MinimumVisualStudioVersion>
     <VsixOutputDirName>VS15</VsixOutputDirName>
     <CodeAnalysisRuleSet Condition="'$(CodeAnalysisRuleSet)' == ''">$(BuildCommonDirectory)NuGet.ruleset</CodeAnalysisRuleSet>
-    <EnforceCodeStyleInBuild Condition=" '$(EnforceCodeStyleInBuild)' != '' " >true</EnforceCodeStyleInBuild>
+    <EnforceCodeStyleInBuild Condition="'$(EnforceCodeStyleInBuild)' == ''">true</EnforceCodeStyleInBuild>
+    <WarningsNotAsErrors>$(WarningsNotAsErrors);EnableGenerateDocumentationFile</WarningsNotAsErrors>
```
Comment (kartheekp-ms) [inline build/common.project.props]: I am curious to learn more about the benefits of `WarningsNotAsErrors` and `EnableGenerateDocumentationFile` properties.

File: build/common.project.props [hunk 5]
```diff
@@ -124,7 +124,8 @@
     <MinimumVisualStudioVersion>15.0</MinimumVisualStudioVersion>
     <VsixOutputDirName>VS15</VsixOutputDirName>
     <CodeAnalysisRuleSet Condition="'$(CodeAnalysisRuleSet)' == ''">$(BuildCommonDirectory)NuGet.ruleset</CodeAnalysisRuleSet>
-    <EnforceCodeStyleInBuild Condition=" '$(EnforceCodeStyleInBuild)' != '' " >true</EnforceCodeStyleInBuild>
+    <EnforceCodeStyleInBuild Condition="'$(EnforceCodeStyleInBuild)' == ''">true</EnforceCodeStyleInBuild>
```
Comment (kartheekp-ms) [inline build/common.project.props:127]: If I am reading the [document](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/overview?tabs=net-8#latest-updates) correctly, we should also set the `AnalysisLevel` MSBuild property to a specific value to lock the warnings to a particular version of SDK. Otherwise, we may get frequent build failures as we keep building on the preview versions of SDK.

Comment (kartheekp-ms) [inline build/common.project.props:127]: If I am reading [document](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/overview?tabs=net-8#enable-on-build) correctly, it looks like we are relying on the default severity set by the .NET SDK Analyzers. I am okay with this but just wanted to share that there is an option to set severity levels of these rules. I think we should start with default severity set by the analyzers and adjust as we progress.

File: build/common.project.props (hunk 4 above)
Comment (jgonz120) [reply build/common.project.props]: EnableGenerateDocumentationFile is a warning that causes the build to fail when EnforceCodeStyleInBuild is enabled and IDE0005 is set to warning or greater. IDE0005 sets a warning for unused import statements. 

So I'm setting EnableGenerateDocumentationFile to not error on compile.

File: build/common.project.props (hunk 5 above)
Comment (jgonz120) [reply build/common.project.props:127]: I like this, at least we can be intentional about taking in new rules as new versions of the SDK are released. Added.

Comment (jgonz120) [reply build/common.project.props:127]: We have an .editorconfig file with a bunch of values already too.

File: build/common.project.props [hunk 6]
```diff
@@ -124,7 +124,9 @@
...
+    <WarningsNotAsErrors>$(WarningsNotAsErrors);EnableGenerateDocumentationFile</WarningsNotAsErrors>
+    <AnalysisLevel>8.0</AnalysisLevel>
```
Comment (nkolev92) [inline build/common.project.props]: I think the spirit of that is to only add it if you need it. 

See details about https://github.com/dotnet/designs/blob/main/proposed/sdk-analysis-level.md, that we collaborated on.

This is a team decision ofc, but I think in general, we want to get a steady stream of warnings, instead of all at once.

Comment (jgonz120) [reply build/common.project.props]: You know, I added this and now we're getting more errors 😅

Comment (jgonz120) [reply build/common.project.props]: Okay so if I run it with level "Latest" it gives a new set of errors.

But the docs say, "If your project targets .NET 5 or later, or if you've added the [AnalysisMode](https://learn.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#analysismode) property, the default value is latest." 

It looks like we target net472 for most of our projects, so I guess that adds up. I'm going to remove it for now bc I don't want to go about fixing all those changes in all those projects 😅

File: build/common.project.props (hunk 4 above)
Comment (jeffkl) [inline build/common.project.props]: Shouldn't this be 

```suggestion
    <WarningsNotAsErrors>$(WarningsNotAsErrors);IDE0005</WarningsNotAsErrors>
```

I see a bunch of warnings on the build https://dev.azure.com/dnceng-public/public/_build/results?buildId=683379&view=results

Comment (jgonz120) [reply build/common.project.props]: That would make it so "Remove unnecessary using directives" won't give an error on build which won't actually enforce that rule. 

But I'm okay with it for this PR since I want to start enabling those types of rules project by project in subsequent PRs.

Comment (jgonz120) [reply build/common.project.props]: Actually second thought. I switched it to NoWarn, that way IDE0005 can still warn for projects as they are enabled.

**Review by nkolev92 — DISMISSED** [review]

File: build/DotNetSdkVersions.txt [hunk 7]
```diff
@@ -1,3 +1,3 @@
 # Each line represents arguments for the .NET SDK installer script (https://learn.microsoft.com/dotnet/core/tools/dotnet-install-script)
--Channel 8.0
+-Channel 8.0.2xx
```
Comment (dtivel) [inline build/DotNetSdkVersions.txt:2]: This seems unrelated to enforcing code style in build.

Can you elaborate on this change?  `8.0` will use the current released version, which is currently 8.0.3xx.  Hard coding this to `8.0.2xx` will lock us into that feature band until someone bumps it.

Comment (jgonz120) [reply build/DotNetSdkVersions.txt:2]: There's a known issue that was causing the build to fail. If we use 8.0.3xx then there is a mismatch between the compiler versions. We will need to update the value back once the latest version is available in CI. 

https://github.com/dotnet/core/blob/21da990bab4170059c6e9ed84feba7e912fa56bc/release-notes/8.0/known-issues.md#802xx-sdk-is-not-compatible-with-178-for-some-scenarios

**Review by donnie-msft — DISMISSED** [review]

**Review by nkolev92 — APPROVED** [review]

File: build/DotNetSdkVersions.txt (hunk 7 above)
Comment (jgonz120) [reply build/DotNetSdkVersions.txt:2]: @dtivel is it okay to resolve this comment?
